### PR TITLE
Allow generation of mfgtool bundle

### PR DESCRIPTION
**Pull request by otavio**
The class provides the infrastructure for MFGTOOL generation and is tied
to images. To generate the bundle, the task populate_mfgtool must be
called. For example:

,----[ Running populate_mfgtool for core-image-minimal image ]
| $: bitbake core-image-minimal -c populate_mfgtool
`----

The class behavior is controlled through the MFGTOOL_SCRIPT variable. The
MFGTOOL_SCRIPT variable itself specifies a space-separated list of the
scripts to enable. Following the scripts, you can determine the behavior
of each script by providing up to two order-dependent arguments, which
are separated by commas. You can omit any argument you like but must
retain the separating commas. The order is important and specifies the
following:

  1. Extra dependencies that should be added to the do_populate_mfgtool
     task, if the script is enabled.
  2. Extra binaries that should be added to the bundle, if the script is
     enabled.

For example:

,----[ Defining bootloader.uuu.in script ]
| MFGTOOL_SCRIPT = "bootloader.uuu.in"
| MFGTOOL_SCRIPT[bootloader.uuu.in] = "virtual/bootloader:do_deploy,${UBOOT_BINARY}"
`----

The virtual/bootloader:do_deploy is added to do_populate_mfgtool
dependencies and ${UBOOT_BINARY} copied to the bundle, only if the
script is enabled.

During the mfgtool bundle generation, the uuu.in files are processed and
some variables replaced. The variables are:

  - MACHINE
  - UBOOT_BINARY
  - SPL_BINARY
  - IMAGE_BASENAME

---

TODO Items before merging:
- [ ] add an i.MX8MM uuu script set
- [ ] add an i.MX6QDL uuu script set

### CODE REVIEW

File: classes/image_populate_mfgtool.bbclass [hunk 1]
```diff
@@ -0,0 +1,142 @@
+# Allow generation of mfgtool bundle
```
Comment (angolini) [inline classes/image_populate_mfgtool.bbclass:1]: there are 2 typos in the commit log:
tow/two
exemple/example

File: classes/image_populate_mfgtool.bbclass [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+# The class provides the infrastructure for MFGTOOL generation and is tied to images. To generate
+# the bundle, the task populate_mfgtool must be called. For exemple:
+#
+# ,----[ Running populate_mfgtool for core-image-minimal image ]
+# | $: bitbake core-image-minimal -c populate_mfgtool
+# `----
+#
+# The class behavior is controlled through the MFGTOOLSCRIPT variable. The MFGTOOLSCRIPT variable
+# itself specifies a space-separated list of the scripts to enable. Following the scripts, you can
+# determine the behavior of each script by providing up to tow order-dependent arguments, which are
```
Comment (angolini) [inline classes/image_populate_mfgtool.bbclass:12]: tow/two

File: classes/image_populate_mfgtool.bbclass [hunk 3]
```diff
@@ -0,0 +1,142 @@
+# Allow generation of mfgtool bundle
+#
+# The class provides the infrastructure for MFGTOOL generation and is tied to images. To generate
+# the bundle, the task populate_mfgtool must be called. For exemple:
```
Comment (angolini) [inline classes/image_populate_mfgtool.bbclass:4]: exemple/example

File: classes/image_populate_mfgtool.bbclass [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+#
+# ,----[ Running populate_mfgtool for core-image-minimal image ]
+# | $: bitbake core-image-minimal -c populate_mfgtool
+# `----
+#
+# The class behavior is controlled through the MFGTOOLSCRIPT variable. The MFGTOOLSCRIPT variable
+# itself specifies a space-separated list of the scripts to enable. Following the scripts, you can
+# determine the behavior of each script by providing up to tow order-dependent arguments, which are
+# separated by commas. You can omit any argument you like but must retain the separating commas. The
```
Comment (angolini) [inline classes/image_populate_mfgtool.bbclass:13]: the arguments are separated by commas. Maybe the list (for each argument) is separated by spaces?

**Review by angolini — COMMENTED** [review]
I like this, let me test on my side to make sure I understand how to use

File: classes/image_populate_mfgtool.bbclass (hunk 4 above)
Comment (otavio) [reply classes/image_populate_mfgtool.bbclass:13]: The use of `,` is important as we can have multi dependencies and files.

Comment (otavio) [reply classes/image_populate_mfgtool.bbclass:13]: I think I improved it.

File: classes/image_populate_mfgtool.bbclass [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+MFGTOOL_FILESPATH ??= " \
+    ${@base_set_filespath(["%s/mfgtool" % p for p in "${BBPATH}".split(":")] \
+                             + ["${FILE_DIRNAME}/${BP}/mfgtool", \
+                                "${FILE_DIRNAME}/${BPN}/mfgtool", \
+                                "${FILE_DIRNAME}/files/mfgtool"] \
+                          , d)} \
+"
+
+MFGTOOLDIR = "${WORKDIR}/mfgtool-${PN}"
```
Comment (quaresmajose) [inline classes/image_populate_mfgtool.bbclass:50]: The `${PN}` is already present on the `${WORKDIR}` of the image so maybe it can be dropped.

File: classes/image_populate_mfgtool.bbclass [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+            if flag in mfgtoolscript:
+                if num >= 2 and items[1]:
+                    mfgtoolscript_deploy_files.append(items[1])
+                if num >= 1 and items[0]:
+                    mfgtoolscript_depends.append(items[0])
+
+        d.appendVarFlag('do_populate_mfgtool', 'depends', ' ' + ' '.join(mfgtoolscript_depends))
+
+    d.setVar('MFGTOOLSCRIPT_DEPLOY_FILES', ' '.join(mfgtoolscript_deploy_files))
```
Comment (quaresmajose) [inline classes/image_populate_mfgtool.bbclass:83]: as `MFGTOOLSCRIPT_DEPLOY_FILES` is an internal variable of the class, something like `_MFGTOOLSCRIPT_DEPLOY_FILES` can be better.

**Review by quaresmajose — COMMENTED** [review]
LGTM

File: classes/image_populate_mfgtool.bbclass [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+#
+# The class provides the infrastructure for MFGTOOL generation and is tied to images. To generate
+# the bundle, the task populate_mfgtool must be called. For example:
+#
+# ,----[ Running populate_mfgtool for core-image-minimal image ]
+# | $: bitbake core-image-minimal -c populate_mfgtool
+# `----
+#
+# The class behavior is controlled through the MFGTOOLSCRIPT variable. The MFGTOOLSCRIPT variable
+# itself specifies a space-separated list of the scripts to enable. Following the scripts, you can
```
Comment (thochstein) [inline classes/image_populate_mfgtool.bbclass:11]: To improve usability, I suggest updating the variable name with `_` and `S`: `MFGTOOL_SCRIPTS`.

File: classes/image_populate_mfgtool.bbclass [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+    for src in $(ls -1 ${WORKDIR}/*.uuu); do
+        dest=$(basename $src)
+        bbnote " - $src -> ${MFGTOOLDIR}/${PN}-${MACHINE}/$dest"
+        install -D -m 0644 $src ${MFGTOOLDIR}/${PN}-${MACHINE}/$dest
+    done
+
+    bbnote "Copying uuu binaries..."
+    cp -v -s ${STAGING_LIBDIR_NATIVE}/uuu/* ${MFGTOOLDIR}/${PN}-${MACHINE}/
+
+    bbnote "Copyng MFGTOOL extra deploy files..."
```
Comment (MrCry0) [inline classes/image_populate_mfgtool.bbclass:131]: `s/Copyng/Copying/`

With that and previous notes from other colleagues - LGTM.

File: classes/image_populate_mfgtool.bbclass (hunk 6 above)
Comment (otavio) [reply classes/image_populate_mfgtool.bbclass:83]: Fixed.

File: classes/image_populate_mfgtool.bbclass (hunk 7 above)
Comment (otavio) [reply classes/image_populate_mfgtool.bbclass]: Fixed.

File: classes/image_populate_mfgtool.bbclass (hunk 8 above)
Comment (otavio) [reply classes/image_populate_mfgtool.bbclass]: Fixed.

File: classes/image_populate_mfgtool.bbclass [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+
+addtask populate_mfgtool after do_image_complete do_unpack before do_deploy
+do_populate_mfgtool[dirs] ?= "${DEPLOY_DIR_IMAGE} ${WORKDIR}"
+do_populate_mfgtool[nostamp] = "1"
+do_populate_mfgtool[recrdeptask] += "do_deploy"
+do_populate_mfgtool[depends] += "uuu-bin-native:do_populate_sysroot"
+
+python () {
+    # Process MFGTOOL_SCRIPTS and handle its fields.
+    mfgtoolscript_flags = d.getVarFlags('MFGTOOL_SCRIPTS') or {}
```
Comment (thochstein) [inline classes/image_populate_mfgtool.bbclass:62]: It helps readability to maintain consistency, so `mfgtoolscript_flags` should be `mfgtool_scripts_flags`.

File: classes/image_populate_mfgtool.bbclass [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+    mfgtoolscript_depends = []
+    mfgtoolscript_deploy_files = []
+    if mfgtoolscript_flags:
+        mfgtoolscript = (d.getVar('MFGTOOL_SCRIPTS') or "").split()
```
Comment (thochstein) [inline classes/image_populate_mfgtool.bbclass:66]: `mfgtoolscript` -> `mfgtool_scripts`.

Comment (otavio) [reply classes/image_populate_mfgtool.bbclass:66]: Fixed.

File: classes/image_populate_mfgtool.bbclass (hunk 9 above)
Comment (otavio) [reply classes/image_populate_mfgtool.bbclass:62]: Fixed.

**Comment by otavio** [comment]
I ended up using singular for the variable name. The main reason for using singular is because every entry sets properties for a _specific_ script. So the dependencies and binaries for installation belong to one particular script.

**Comment by thochstein** [comment]
The variable `MFGTOOL_SCRIPTS` represents a _set of scripts_ and so should be plural. The variable `MFGTOOL_SCRIPTS[script]` is itself singular, representing the configuration for the single member named `script`.

File: classes/image_populate_mfgtool.bbclass [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+                    script_depends.append(items[0])
+
+        d.appendVarFlag('do_populate_mfgtool', 'depends', ' ' + ' '.join(script_depends))
+
+    d.setVar('_SCRIPT_DEPLOY_FILES', ' '.join(script_deploy_files))
+}
+
+python do_populate_mfgtool() {
+    # For MFGTOOL_SCRIPT items we use BitBake's fetcher module allowing a consistent behavior.
+    script = (d.getVar('MFGTOOL_SCRIPT') or "").split()
```
Comment (thochstein) [inline classes/image_populate_mfgtool.bbclass:88]: Consider if you wanted to do a `for-loop`, best would be `for script in scripts`.

File: classes/image_populate_mfgtool.bbclass [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+do_populate_mfgtool[dirs] ?= "${DEPLOY_DIR_IMAGE} ${WORKDIR}"
+do_populate_mfgtool[nostamp] = "1"
+do_populate_mfgtool[recrdeptask] += "do_deploy"
+do_populate_mfgtool[depends] += "uuu-bin-native:do_populate_sysroot"
+
+python () {
+    # Process MFGTOOL_SCRIPT and handle its fields.
+    script_flags = d.getVarFlags('MFGTOOL_SCRIPT') or {}
+    script_depends = []
+    script_deploy_files = []
```
Comment (thochstein) [inline classes/image_populate_mfgtool.bbclass:64]: I see that this is where naming is tricky. The problem is `script_depends` and `script_deploy_files` here are sets of sets, so it's hard to come up with an accurate name that is not cumbersome. Although it could be argued that `script_depends` would be clearly a single set for a single `script`, while `scripts_depends` would be the full 'set of sets'.

How about using `flags`, `depends`, and `deploy_files` here, and `script_depends` and `script_deploy_files` for a single `script`?

File: classes/image_populate_mfgtool.bbclass (hunk 11 above)
Comment (otavio) [reply classes/image_populate_mfgtool.bbclass:88]: Sure, but the `MFGTOOL_SCRIPT` is analogous to `PACKAGECONFIG`. It defines a script to have flags (depends and files) defined.

Comment (thochstein) [reply classes/image_populate_mfgtool.bbclass:88]: Let me try again: I consider that `PACKAGECONFIG` has a special definition, that a configuration is a set of values defining the singular configuration. I see no such definition for `MFGTOOL_SCRIPT`. It is not a set of values defining a singular script, so it is a bad analog.

Comment (thochstein) [reply classes/image_populate_mfgtool.bbclass:88]: Hmm, how about using `MFGTOOLCONFIG`? I guess that makes your analog clearer to me.

Comment (otavio) [reply classes/image_populate_mfgtool.bbclass:88]: `MFGTOOL_SCRIPT`: I consider it defines a script with a set of values. Consider this:

```
MFGTOOL_SCRIPT = "foo.uuu.in bar.uuu"
MFGTOOL_SCRIPT[foo.uuu.in] = "dep-foo1:do_deploy dep-foo2:do_deploy,file-foo1 file-foo2"
MFGTOOL_SCRIPT[bar.uuu] = "dep-bar1:do_deploy dep-bar2:do_deploy,file-bar1 file-bar2"
```

This defines two scripts, and each has a specific set of values. A convention is to use the same variable (here `MFGTOOL_SCRIPT`) to look up the flags.

Comment (thochstein) [reply classes/image_populate_mfgtool.bbclass:88]: The question is whether the variable name can be improved. Calling it `MFGTOOL_SCRIPT` gives a natural connotation that only one script is allowed. A knowing user will go to the docs to see if multiples are allowed. Making it plural makes it clear from the name.

As I said, I think all this goes away if you just make the variable `MFGTOOL_CONFIG`, aligning it with `PACKAGECONFIG` and `UBOOT_CONFIG`, so giving the user another clue about how it works just from the name.

**Review by thochstein — APPROVED** [review]

**Comment by thochstein** [comment]
Thanks :-)

File: classes/image_populate_mfgtool.bbclass (hunk 11 above)
Comment (otavio) [reply classes/image_populate_mfgtool.bbclass]: Fixed.

**Review by MrCry0 — APPROVED** [review]

File: recipes-devtools/uuu/uuu-bin-native_1.4.243.bb [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+    install -D -m 0755 ${WORKDIR}/uuu-${PV}_mac ${D}${libdir}/uuu/uuu_mac
+    install -D -m 0644 ${WORKDIR}/uuu-${PV}.exe ${D}${libdir}/uuu/uuu.exe
+}
+
+SYSROOT_DIRS_NATIVE = "${libdir}/uuu"
+
+# HACK! We are not aiming to run those binaries during the build but copy then for MFGTOOL
+# bundle. If make the binaries relocatable they will rely on OpenEmbedded native sysroot paths but
+# we want them to run at user machine.
+SYSROOT_PREPROCESS_FUNCS:remove = "relocatable_binaries_preprocess relocatable_native_pcfiles"
```
Comment (quaresmajose) [inline recipes-devtools/uuu/uuu-bin-native_1.4.243.bb:37]: @otavio maybe is better to use only the `do_deploy` with `INHIBIT_DEFAULT_DEPS = "1"` for this propose. with that, the `inherit native` can also be removed.

Comment (otavio) [reply recipes-devtools/uuu/uuu-bin-native_1.4.243.bb]: I changed the way it works, and it looks good to me. Please take a look.

Comment (quaresmajose) [reply recipes-devtools/uuu/uuu-bin-native_1.4.243.bb]: even better deploy the archive with everything inside.

**Review by quaresmajose — APPROVED** [review]
LGTM

**Review by kraj — APPROVED** [review]
lgtm
